packages/framework/src/client/scheduler.js

```diff
@@ -0,0 +1,203 @@
+import debugFactory from 'debug';
+import { find } from 'lodash';
+import ResourceRequest, { STATUS } from './resource-request';
+
+const debug = debugFactory( 'fresh-data:scheduler' );
+
+export default class Scheduler {
+	constructor(
+		fetch,
+		setTimeout = window.setTimeout,
+		clearTimeout = window.clearTimeout,
+	) {
+		this.fetch = fetch;
+		this.requests = [];
+		this.setTimeout = setTimeout;
+		this.clearTimeout = clearTimeout;
+		this.timeoutId = null;
+		this.nextRequestTime = null;
+	}
+
+	/**
+	 * Cancels the processing of requests.
+	 */
+	stop = () => {
+		if ( null !== this.timeoutId ) {
+			debug( 'Cancelling next update.' );
+			this.clearTimeout( this.timeoutId );
+			this.timeoutId = null;
+			this.nextRequestTime = null;
+		}
+	}
+
+	/**
+	 * (Re)schedules next run of request processing, based on the current collection of requests.
+	 * Note: Only reschedules if necessary.
+	 * @param {Date} now The current time
+	 */
+	updateDelay = ( now = new Date() ) => {
+		const currentDelay = this.nextRequestTime ? now.getTime() - this.nextRequestTime : null;
+		const nextDelay = this.getNextRequestDelay( now );
+
+		if ( nextDelay !== currentDelay ) {
```

Comment:
This logic seems flaky because `currentDelay` is calculated before doing some computations. If those computations take more than `1ms`, `nextDelay` will be different than `currentDelay` even if the scheduled time hasn't changed. Did you find any problem when testing this?

I can think of several options to solve this:
- Always re-schedule. Calling `clearTimeout` and `setTimeout` is not an expensive operation, specially since you're not creating new Functions. That would simplify the `updateDelay` logic a bit.
- Instead of the `ResourceRequest.getTimeLeft()` returning the amount of time from `now`, it could return an absolute timestamp.
- Have some tolerance margin, like this:
```suggestion
		if ( Math.abs( nextDelay - currentDelay ) > 50 ) {
```

But all this is just a nitpick, the `nextDelay !== currentDelay` is just an optimization so it's ok if it fails sometimes. Fix it if you feel like it :)

Reply:
I like the always reschedule option here, any time I can simplify and reduce code without major performance hits, I'm all for it. I'll look into that.

Reply:
Updated in f44b57c0276d2b7125e701eed713f9f35dd57f55